src/mon/MgrMonitor.cc

```diff
@@ -0,0 +1,241 @@
+  pending_map.epoch++;
+}
+
+void MgrMonitor::encode_pending(MonitorDBStore::TransactionRef t)
+{
+  bufferlist bl;
+  pending_map.encode(bl, 0);
+  put_version(t, pending_map.epoch, bl);
+  put_last_committed(t, pending_map.epoch);
+}
+
+bool MgrMonitor::preprocess_query(MonOpRequestRef op)
+{
+  PaxosServiceMessage *m = static_cast<PaxosServiceMessage*>(op->get_req());
+  switch (m->get_type()) {
+    case MSG_MGR_BEACON:
+      return preprocess_beacon(op);
+    default:
+      mon->no_reply(op);
+      derr << "Unhandled message type " << m->get_type() << dendl;
+      return true;
+  }
+}
+
+bool MgrMonitor::prepare_update(MonOpRequestRef op)
+{
+  PaxosServiceMessage *m = static_cast<PaxosServiceMessage*>(op->get_req());
+  switch (m->get_type()) {
+    case MSG_MGR_BEACON:
+      return prepare_beacon(op);
+    default:
+      mon->no_reply(op);
+      derr << "Unhandled message type " << m->get_type() << dendl;
+      return true;
+  }
+}
+
+bool MgrMonitor::preprocess_beacon(MonOpRequestRef op)
+{
+  //MMgrBeacon *m = static_cast<MMgrBeacon*>(op->get_req());
+
+  return false;
+}
+
+class C_Updated : public Context {
+  MgrMonitor *mm;
+  MonOpRequestRef op;
+public:
+  C_Updated(MgrMonitor *a, MonOpRequestRef c) :
+    mm(a), op(c) {}
+  void finish(int r) {
+    if (r >= 0) {
+      // Success 
+    } else if (r == -ECANCELED) {
+      mm->mon->no_reply(op);
+    } else {
+      mm->dispatch(op);        // try again
+    }
+  }
+};
+
+bool MgrMonitor::prepare_beacon(MonOpRequestRef op)
+{
+  MMgrBeacon *m = static_cast<MMgrBeacon*>(op->get_req());
+
+  pending_map.active_gid = m->get_gid();
+  pending_map.active_addr = m->get_server_addr();
+
+  dout(4) << "proposing epoch " << pending_map.epoch << dendl;
+  wait_for_finished_proposal(op, new C_Updated(this, op));
+
+  return true;
+}
+
+void MgrMonitor::check_subs()
+{
+  const std::string type = "mgrmap";
+  if (mon->session_map.subs.count(type) == 0)
+    return;
+  xlist<Subscription*>::iterator p = mon->session_map.subs[type]->begin();
+  while (!p.end()) {
+    Subscription *sub = *p;
+    ++p;
+    check_sub(sub);
+  }
+}
+
+void MgrMonitor::check_sub(Subscription *sub)
+{
+  if (sub->type == "mgrmap") {
+    if (sub->next <= map.get_epoch()) {
+      dout(20) << "Sending map to subscriber " << sub->session->con << dendl;
+      sub->session->con->send_message(new MMgrMap(map));
+      if (sub->onetime) {
+        mon->session_map.remove_sub(sub);
+      } else {
+        sub->next = map.get_epoch() + 1;
+      }
+    }
+  } else {
+    assert(sub->type == "mgrdigest");
+    send_digests();
+  }
+}
+
+/**
+ * Handle digest subscriptions separately (outside of check_sub) because
+ * they are going to be periodic rather than version-driven.
+ */
+void MgrMonitor::send_digests()
+{
+  const std::string type = "mgrdigest";
+  if (mon->session_map.subs.count(type) == 0)
+    return;
+  xlist<Subscription*>::iterator p = mon->session_map.subs[type]->begin();
+  while (!p.end()) {
```

Comment:
@jcsp, see http://en.cppreference.com/w/cpp/language/range-for, and `xlist` does have `begin()` and `end()`, and a proper iterator implementation. see fc9426a5a11e93bb17bbb944b2555fa6ac4183f3. so i guess it supports range based for loop?